Stop adding accessors to their parent DeclContext

slavapestov opened the pull request:
This will help with request-based accessor synthesis. It should also make it easier to lift the restriction on local variables not being allowed to use ‘lazy’ or property wrappers.

slavapestov committed:
SIL: VTable visitor now visits accessors as part of their storage

This is NFC for now, but soon accessors will not no longer appear
when iterating over NominalTypeDecl::getMembers().

slavapestov committed:
AST: Accessors no longer appear as members of their parent DeclContext

Accessors logically belong to their storage and can be synthesized
on the fly, so removing them from the members list eliminates one
source of mutability (but doesn't eliminate it; there are also
witnesses for derived conformances, and implicit constructors).

Since a few ASTWalker implementations break in non-trivial ways when
the traversal is changed to visit accessors as children of the storage
rather than peers, I hacked up the ASTWalker to optionally preserve
the old traversal order for now. This is ugly and needs to be cleaned up,
but I want to avoid breaking _too_ much with this commit.

slavapestov committed:
AST: Clean up findOverridingDecl() and findImplementingMethod()

These are two slightly different operations. Make them share code and
prepare them for when we stop adding accessors to their parent
DeclContext.

Note the FIXMEs; this code could use additional attention.

slavapestov committed:
ASTPrinter: Print details about MissingMemberDecls for debugging purposes

slavapestov committed:
Serialization: Track vtable slots for VarDecl and SubscriptDecl

Once accessors are no longer listed as members of their parent context,
a failure to deserialize a VarDecl or SubscriptDecl needs to create a
MissingMemberDecl with the total number of vtable entries expected for
all of the accessors of the storage.

Note that until the accessor change actually lands, we always compute
the expected number of vtable entries as 0.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

rjmccall commented on lib/AST/ASTWalker.cpp:
```diff
@@ -146,6 +146,19 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
+              return true;
+          }
+        }
+      }
```
Consider extracting a helper function for this so that it doesn't have to be repeated in every DC.

slavapestov replied:
This is a compatibility behavior that's going away soon.

rjmccall commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -1019,24 +1019,31 @@ class SILGenType : public TypeMemberVisitor<SILGenType> {
+
+  void visitAccessors(AbstractStorageDecl *asd) {
+    for (auto *accessor : asd->getAllAccessors())
+      visitFuncDecl(accessor);
```
We shouldn't be introducing new uses of `getAllAccessors` like this, right?  Is it important to visit non-expected accessors here?

slavapestov replied:
I'm planning on refactoring all the calls shortly. For now, I'm just preserving the old behavior in a few places to minimize moving parts.

rjmccall commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -1153,11 +1163,17 @@ class SILGenExtension : public TypeMemberVisitor<SILGenExtension> {
+
+  void visitAccessors(AbstractStorageDecl *asd) {
+    for (auto *accessor : asd->getAllAccessors())
+      visitFuncDecl(accessor);
```
Same question.

rjmccall commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2314,6 +2314,11 @@ class AvailabilityWalker : public ASTWalker {
+  bool shouldWalkAccessorsTheOldWay() override {
+    return true;
+  }
+
```
I'm surprised at this one; why do we need to check availability at all on an accessor?  Or is this checking the function body?

slavapestov replied:
We visit accessors regardless of this method returning true or false. The difference is whether we visit them as children or peers of the storage. I need to tweak the few visitors that override this method to work with the new ordering, then I can remove the legacy code path altogether.

rjmccall replied:
Yeah, I understand the basics of why you've introduced this method, I just figured you'd seen actual problems at each of the sites where you did this.  Doesn't really matter.

slavapestov replied:
Yeah, there was a problem with availability checking of accessors that I haven't investigated yet. I don't expect it will be difficult to resolve.

rjmccall commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2603,6 +2603,10 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+
+    // Now check all the accessors.
+    for (auto *accessor : VD->getAllAccessors())
+      visit(accessor);
```
Same comment about `getAllAccessors`, although it seems even more pertinent here because of the order-sensitivity of which accessors will exist at this point.

slavapestov replied:
Yeah, eventually typeCheckDecl() will only be called on parsed accessors.

slavapestov commented:
@rjmccall Thanks for the review. Are you OK with this getting merged in its current state (once I fix the source compat failures) given that I'm planning on continuing the refactoring? I just didn't want to change too much all at once and break everything, hence the hacks here and there.

rjmccall commented:
> @rjmccall Thanks for the review. Are you OK with this getting merged in its current state (once I fix the source compat failures) given that I'm planning on continuing the refactoring? I just didn't want to change too much all at once and break everything, hence the hacks here and there.

I'm fine with you committing this as an incremental step, yeah.  Just a little curious that you chose to make incremental process by introducing new dependencies on a somewhat questionable method.

slavapestov commented:
@rjmccall We don't do these things because they're easy.

slavapestov committed:
SILOptimizer: Fix speculative devirtualization of constructor calls

By unintentionally using SILDeclRef::getFuncDecl() we were getting a nullptr back,
and passing it in to ClassDecl::getOverridingDecl(). This started crashing with a
refactoring I'm making. Fix this and add a test.

slavapestov commented:
https://github.com/apple/swift-lldb/pull/1833
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please test Linux

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -3179,6 +3179,12 @@ void PrintAST::visitModuleDecl(ModuleDecl *decl) { }
   recordDeclLoc(decl, [&]{ Printer << decl->getFullName(); });
+  unsigned numVTableEntries = decl->getNumberOfVTableEntries();
+  if (numVTableEntries > 0)
+    Printer << " (vtable entries: " << numVTableEntries << ")";
```
You were going to put this behind an option, no?

slavapestov replied:
It appears that SkipMissingMemberPlaceholders is true by default, so we should only be printing these from swift-ide-test. Do you still want an option?

jrose-apple replied:
Ah, didn't realize that was already there! Okay, yeah, this is fine.

jrose-apple commented on include/swift/AST/ASTWalker.h:
```diff
@@ -164,7 +164,7 @@ class ASTWalker {
 
   /// walkToDeclPre - This method is called when first visiting a decl, before
-  /// walking into its children.  If it returns false, the subtree is skipped.
+  /// walking into its children.  If it returns true, the subtree is skipped.
```
Are you sure about these? That seems wrong when the default behavior is to return `true`.

slavapestov replied:
I... don't know what I was thinking. You're absolutely right, this makes no sense. I'll remove this commit.

jrose-apple commented on include/swift/Serialization/ModuleFormat.h:
```diff
@@ -1064,6 +1064,7 @@ namespace decls_block {
     AccessLevelField, // setter access, if applicable
     DeclIDField, // opaque return type decl
     BCFixed<2>,  // # of property wrapper backing properties
+    BCFixed<8>, // total number of vtable entries introduced by all accessors
```
`BCVBR<4>` recommended

slavapestov replied:
I remember you mentioned that and I had forgotten about it. Thanks for the reminder!

jrose-apple reviewed (commented):
I was wondering whether you want to take *out* the `needs new vtable entry` field from AccessorLayout, but I guess it's okay to leave it, especially if we ever need to support "the getter could be loaded but not the setter" for some reason.

slavapestov commented:
@jrose-apple The 'needs new vtable entry' in AccessorDecl is still needed for vtable layout. We 'flatten' storage decls into accessors in SILVTableVisitor and add a separate vtable entry (or not) for each individual accessor.

jrose-apple commented:
Oh, right, of course; it's not just used for MissingMemberDecl. Okay, makes sense.

slavapestov commented:
https://github.com/apple/swift/pull/26356
@swift-ci Please smoke test

slavapestov commented:
https://github.com/apple/swift-lldb/pull/1833
@swift-ci Please smoke test
